### PR TITLE
tests/robustness: Implement Kubernetes pagination

### PR DESCRIPTION
**Pull request by serathius**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by serathius** [comment]
cc @ahrtr @ptabor @jmhbnz @chaochn47 @fuweid

File: tests/robustness/traffic/kubernetes.go [hunk 1]
```diff
@@ -102,54 +94,92 @@ func (t kubernetesTraffic) Run(ctx context.Context, c *RecordingClient, limiter
...
 			}
-			limiter.Wait(ctx)
 		}
 	})
 	g.Wait()
 }
 
-func (t kubernetesTraffic) Write(ctx context.Context, kc *kubernetesClient, ids identity.Provider, s *storage) (err error) {
+func (t kubernetesTraffic) Read(ctx context.Context, kc *kubernetesClient, s *storage, limiter *rate.Limiter, keyPrefix string) (rev int64, err error) {
+	limit := int64(t.averageKeyCount)
```
Comment (chaochn47) [inline tests/robustness/traffic/kubernetes.go:113]: umm, can we start with a smaller limit to guarantee the following pagination?

File: tests/robustness/traffic/kubernetes.go [hunk 2]
```diff
@@ -102,54 +94,92 @@ func (t kubernetesTraffic) Run(ctx context.Context, c *RecordingClient, limiter
...
+	}
+	limiter.Wait(ctx)
+	hasMore := resp.More
+	revision = resp.Header.Revision
+	kvs := resp.Kvs
+
+	for hasMore {
+		lastKey := string(kvs[len(kvs)-1].Key)
+		readCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
+		resp, err := kc.Range(readCtx, lastKey+"\x00", keyPrefix+"\xff", revision, limit)
```
Comment (chaochn47) [inline tests/robustness/traffic/kubernetes.go:129]: `keyPrefix+"\xff` can be replaced with `clientv3.GetPrefixRangeEnd(prefix)`

File: tests/robustness/traffic/kubernetes.go (hunk 1 above)
Comment (serathius) [reply tests/robustness/traffic/kubernetes.go:113]: Based on experiments this is enough to guarantee pagination, any less there is too much requests.

File: tests/robustness/traffic/kubernetes.go [hunk 3]
```diff
@@ -168,14 +198,33 @@ type kubernetesClient struct {
...
+	copy(end, key)
+	for i := len(end) - 1; i >= 0; i-- {
+		if end[i] < 0xff {
+			end[i] = end[i] + 1
+			end = end[:i+1]
+			return string(end)
+		}
+	}
+	return "\x00"
+}
```
[the remark this answers is not in the document]
Comment (serathius) [reply tests/robustness/traffic/kubernetes.go:222]: Looks good, must have missed that function.

File: tests/robustness/traffic/kubernetes.go (hunk 2 above)
Comment (serathius) [reply tests/robustness/traffic/kubernetes.go:129]: Done

**Review by chaochn47 — APPROVED** [review]
LGTM, thanks @serathius!

File: tests/robustness/traffic/kubernetes.go (hunk 1 above)
Comment (chaochn47) [reply tests/robustness/traffic/kubernetes.go:113]: > We get one range request per two lists.

I am not sure I understand this statement. I would expect multiple range requests for each list. This should fall into the second goroutine. 

First `Read` is used in the first goroutine which is ListAndWatch in kubernets. It's a full scan of the resource key prefix range. 

The second `Read` in another goroutine is k8s paginated list requests, each page count is 500 in kubernetes by default. Usually it results in client side pagination.

**Pull request merged**